src/libraries/System.Net.Sockets/tests/FunctionalTests/NetworkStreamTest.cs

```diff
@@ -321,21 +321,31 @@ public async Task DisposeSocketDirectly_ReadWriteThrowNetworkException(bool deri
                 Task<Socket> acceptTask = listener.AcceptAsync();
                 await Task.WhenAll(acceptTask, client.ConnectAsync(new IPEndPoint(IPAddress.Loopback, ((IPEndPoint)listener.LocalEndPoint).Port)));
                 using Socket serverSocket = await acceptTask;
+
                 using NetworkStream server = derivedNetworkStream ? (NetworkStream)new DerivedNetworkStream(serverSocket) : new NetworkStream(serverSocket);
-                
+
                 serverSocket.Dispose();
 
-                Assert.Throws<IOException>(() => server.Read(new byte[1], 0, 1));
-                Assert.Throws<IOException>(() => server.Write(new byte[1], 0, 1));
+                ExpectIOException(() => server.Read(new byte[1], 0, 1), isWriteOperation: false);
+                ExpectIOException(() => server.Write(new byte[1], 0, 1), isWriteOperation: true);
 
-                Assert.Throws<IOException>(() => server.Read((Span<byte>)new byte[1]));
-                Assert.Throws<IOException>(() => server.Write((ReadOnlySpan<byte>)new byte[1]));
+                ExpectIOException(() => server.Read((Span<byte>)new byte[1]), isWriteOperation: false);
+                ExpectIOException(() => server.Write((ReadOnlySpan<byte>)new byte[1]), isWriteOperation: true);
 
-                Assert.Throws<IOException>(() => server.BeginRead(new byte[1], 0, 1, null, null));
-                Assert.Throws<IOException>(() => server.BeginWrite(new byte[1], 0, 1, null, null));
+                ExpectIOException(() => server.BeginRead(new byte[1], 0, 1, null, null), isWriteOperation: false);
+                ExpectIOException(() => server.BeginWrite(new byte[1], 0, 1, null, null), isWriteOperation: true);
 
-                Assert.Throws<IOException>(() => { server.ReadAsync(new byte[1], 0, 1); });
-                Assert.Throws<IOException>(() => { server.WriteAsync(new byte[1], 0, 1); });
+                ExpectIOException(() => { _ = server.ReadAsync(new byte[1], 0, 1); }, isWriteOperation: false);
+                ExpectIOException(() => { _ = server.WriteAsync(new byte[1], 0, 1); }, isWriteOperation: true);
+            }
+
+            static void ExpectIOException(Action action, bool isWriteOperation)
+            {
+                IOException ex = Assert.Throws<IOException>(action);
+                string expectedSubstring = isWriteOperation ? "write data" : "read data";
+                Assert.Contains(expectedSubstring, ex.Message);
+                Assert.NotNull(ex.InnerException);
```

Comment:
This NotNull check isn't necessary.  If it's null, it'll fail the IsType check below.